### PR TITLE
 use wp_kses to sanitize svg

### LINKED ISSUE #84: Add a message about securing file uploads
Hi there,

I think it would be a great addition to include the following message below the file upload types table.  

> While allowing various file uploads can be convenient, it poses security risks.
Enhance your site's security by restricting uploads to specific, safe file types and limiting upload permissions to trusted users only.  
Please refer [here](https://wpforms.com/docs/how-to-allow-additional-file-upload-types/?utm_source=WordPress&utm_medium=fut-help&utm_campaign=file-upload-types&utm_content=view-documentation#secure-upload) for more information.

Probably in a lighter color than the heading!  

Thanks 🍻

### PR DESCRIPTION
**Pull request by kkarpieszuk**
Fixes #84

### CODE REVIEW

File: src/Sanitizer.php [hunk 1]
```diff
@@ -126,8 +132,21 @@ private function sanitize_content( string $content ) {
...
 		}
 
-		$content = $this->remove_php_tags( $content );
-		$content = $this->remove_js_tags( $content );
+		if ( $type === 'svg' ) {
+			$allowed_tags = $this->get_allowed_tags_for_svg();
+
+			$xml_content = '';
+
+			if ( strpos( $content, '<?xml' ) === 0 ) {
```
Comment (kkarpieszuk) [inline src/Sanitizer.php]: wp_kses removes `<?xml ... ?>` tag, preserve it and add it back after cleanup.

File: src/Sanitizer.php [hunk 2]
```diff
@@ -126,8 +132,21 @@ private function sanitize_content( string $content ) {
...
+				$xml_content = substr( $content, 0, strpos( $content, '?>' ) + 2 );
+			}
+
+			$content = $this->remove_js_tags( $content );
```
Comment (kkarpieszuk) [inline src/Sanitizer.php]: wp_kses has a bug that when <script> contains code wrapped into `// <![CDATA[` old mosaic trick it incorrectly strips this tag, file:

```
<?xml version="1.0" encoding="utf-8"?><!-- Uploaded to: SVG Repo, www.svgrepo.com, Generator: SVG Repo Mixer Tools -->
<svg
  viewBox="0 0 10 10"
  height="120px"
  width="120px"
  xmlns="http://www.w3.org/2000/svg">
  <circle cx="5" cy="5" r="4" />

  <script>
    // <![CDATA[
    function getColor() {
      const R = Math.round(Math.random() * 255)
        .toString(16)
        .padStart(2, "0");

      const G = Math.round(Math.random() * 255)
        .toString(16)
        .padStart(2, "0");

      const B = Math.round(Math.random() * 255)
        .toString(16)
        .padStart(2, "0");

      return `#${R}${G}${B}`;
    }

    document.querySelector("circle").addEventListener("click", (e) => {
      e.target.style.fill = getColor();
    });
    // ]]>
  </script>
</svg>
```

will become 
```
<!-- Uploaded to: SVG Repo, www.svgrepo.com, Generator: SVG Repo Mixer Tools -->
<svg viewBox="0 0 10 10" height="120px" width="120px" xmlns="http://www.w3.org/2000/svg">
  <circle cx="5" cy="5" r="4" />

  
    //  {
      e.target.style.fill = getColor();
    });
    // ]]&gt;
  
</svg>
```
I am keeping this regex method as a last (or more accurate: first) straw to try not to have svg images broken after the upload.

File: src/Sanitizer.php [hunk 3]
```diff
@@ -213,4 +290,35 @@ private function remove_js_tags( string $content ): string {
...
+	 *
+	 * @since 1.5.0
+	 *
+	 * @param string $content File content.
+	 *
+	 * @return string
+	 */
+	private function remove_scripts_from_html( string $content ): string {
+
+		$dom = new DOMDocument();
```
Comment (kkarpieszuk) [inline src/Sanitizer.php]: to still have SVG file not be destroyed by sanitizer, I had to whitelist all elements and attributes except `script` above. 

I should to the same for HTML files but list of elements and attributes is simply too huge. I tried to base on `wp_kses` allowed list for post content + elements like `html`, `head`, `body`, `meta` but still was making HTML files changed to non acceptable level. 

So here I did other approach. Blacklisted script tag but delegated the logic to battlefield tested DomDocument library. It traverses dom document and if some element is a script, it removes it.  


(Tbh we could do the same for SVG but use `loadXML`, not `loadHTML` and we will not have to do those tricks to preserve `<?xml` and overcoming wp_kses bug with CDATA

Comment (robertrofam) [reply src/Sanitizer.php]: There are many ways for XSS payloads (e.g. javascript) to get injected. A personal favorite is `` which will likely bypass this "remove_scripts_from_html()" function.

You may also review the following URL for many more example XSS payloads https://github.com/payloadbox/xss-payload-list

This is why it is important to look into allow-lists, wp_kses can do this.

You should be able to use an AI agent to help save you time building the list of allowable HTML tags and attributes. Both antropic and google gemini should give you accurate answers.

**Review by robertrofam — CHANGES_REQUESTED** [review]
requesting the usage of allow lists

**Comment by kkarpieszuk** [comment]
thanks @robertrofam I changed to whitelists for SVG and HTML. Could you re-check?

File: src/Sanitizer.php [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 namespace FileUploadTypes;
 
+use SimplePie\Sanitize;
```
Comment (kagg-design) [inline src/Sanitizer.php]: I do not see any usage of it. Is it a leftover?

**Review by kagg-design — CHANGES_REQUESTED** [review]

File: src/Sanitizer.php (hunk 4 above)
Comment (kkarpieszuk) [reply src/Sanitizer.php]: yes, removed. 

also removed iframe from whitelist for html tags

**Review by robertrofam — APPROVED** [review]
Looks like it cleans up any unsafe HTML/Tags now. Approved.

**Review by kagg-design — APPROVED** [review]
Thank you, @kkarpieszuk, LGTM!

**Comment by olesia-zhykina** [comment]
Good job @kkarpieszuk!
I rechecked changes related to SVG file uploading on different WP and WPForms pages and didn't find issues. 
The following areas were covered:

- File Media Upload using: Media Uploader, Block Editor, Classic Editor, Gutenberg, Widgets (regular and legacy) for old themes, FSE for modern themes, Divi Builder, Elementor Builder
- on Chrome, Safari browsers
- WP 6.6.2 + PHP 8.3, WP 5.5 + PHP 7.0
- svg, html files with `<script>` code
- WPForms: form with FU filed with Classic, Modern view, with enabled, disabled storing in Media Library
- with enabled / disabled spam verification
 - checked uploaded files in Entry in Entry Details

**Pull request merged**